### DISCUSSION TITLE
LookAndFeel_V4 implementation

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#596] [General]
Hi, 
I'm working on the Juce LookAndFeel_V4 implementation for Ctrlr that has been done halfway on the current release.
I'm almost done but I'm not that sure it looks better. 
I can select either type of LookAndFeel style I want from the Property Pane at the bottom (V1 V2 V3 V4 with light midnight dark and grey schemes) and it updates the canvas right away now.

Here is a preview of the different types of component in V4: 

![Capture d’écran, le 2024-01-15 à 18 59 36](https://github.com/RomanKubiak/ctrlr/assets/55164385/817f0d8f-b3da-4b3a-affa-d043d71220d4)

I'll need to do custom colourSchemes to really get the spirit of the older Ctrlr version with the shades of blue and add it to the colourScheme property combo.

I still struggle to have the popup on right-clicking the panel canvas, it doesn't update until I restart the program, I can't get it to catch the LookAndFeelChanged() declaration.

I'm currently working on adding the colour picker popup when a color property is clicked just like in the Juce demorunner. It's a great addition.

I need to think about a good way to make it downward compatible without showing V4 components on older panels.

If you have any suggestions about all that styling thing, don't hesitate ;)

Damien

**Comment by Ted-juh** [comment]
Sounds promising, really does. I still think it's weird you have to apply the l&f on the whole panel instead of per component. I think it's doable in Juce. But yes, the way the whole l&f class is implemented in Ctrlr is a little 'awkward'. It works but, to me, it's not easy to extend it.

While you are looking at the colour picker, is it doable to add color swatches?  A set of clickable colours which you could change to your liking. 

I wish I could find out myself. But I started looking into adding Juce 7, a project once started by Mr Dnaldoog. I started with 2200 errors and was down to 150 errors. I noticed I was changing things in the boost files to get rid of errors. Which gave me the thought: What if I update Lua (5.4.6) as well? Ditch boost/ luabind which both are not being maintained anymore. Instead get it to work with Sol2 (version 3)  to do the Lua binding. All in all, it means reinventing the wheel. It means there would be separate versions of Ctrlr. But it's just to see if it's doable to get it up to date with today's standards.

Don't get your hopes up of anything usable, it's just to see how far I can get

**Reply by damiensellier** [reply]
I thought about doing it per component too, or sending the lookandfeelchanged from the global panel property and letting the user select the L&F per component with an added drop down menu like the custom colours overrides. This is why I'am giving me time to think about the best way to do it.
Colour picker is already implemented as a class in JUCE doesn't have the color swatch. People talked about it on the JUCE forum to have the colour swatch showing the different tones related to their colourID which is useful. For the moment it's not available unfortunately.

Upgrading to JUCE 7 will be a big pain because 1 line of code to check is worth 5 min of build time so implementing another C++ to LUA library will be a torture for the next 2 years :) I think having a stable version of JUCE 6 with L&F V4 and the VST3 export working properly will be a really nice first step.
I'll add the right click thing to the component area of the customcomponent as you mentionned before. Also I'll add the open panel property in the rightclick menu to the canvas. that's easy to do.

**Reply by damiensellier** [reply]
Yep I think it's a good idea to be able to override the L&F at the component level so I added a selector for that. People will be able to have ugly 1998 3D bubble buttons with flat design ilegible sliders :)

![Capture d’écran, le 2024-01-17 à 00 38 21](https://github.com/RomanKubiak/ctrlr/assets/55164385/066f882c-3898-4527-8f27-3240b11756e0)

**Reply by Ted-juh** [reply]
I have to dig deep to find the right Jucer file but I once implemented the swatches in Juce, Could be the same in Ctrlr then, right?

Just to add, as an addition for us design luvvers that do get itchy when seeing the atrocious flat designs Juce drops on us but we can apply a custom L&F by putting in an init file:

```
lf1 = LookAndFeel()
lf1:setImplementation (slidLF1)
panel:getComponent("slid1"):setCustomLookAndFeel(lf1)
```

and then create an L&F by adding a method, in this case, called slidLF1.
```
slidLF1 = {drawRotarySlider = function(g,x,y,width,height,sliderPosProportional,rotaryStartAngle,rotaryEndAngle,slider)
//Do a paint routine for a rotary slider
end,
}
```

So, if you create a script called slidLF1, slidLF2, slidLF3, and so on, you can attach different L&F to different sliders.  Example: This is done on a rotary slider with custom L&F:
![slid](https://github.com/RomanKubiak/ctrlr/assets/71118185/9b41fcb5-d349-4f49-9d82-5db9985abd39)

**Reply by bijlevel** [reply]
I first thought the L&F V4 was looking good, but I agree with the both of you, it's not. It's too flat! @Ted-juh Your custom L&F slider looks way better, but by using KnobMan one can achieve also nice results, at least I think so. This is a partly screenshot of the panel I'm working on with knobs and sliders made with KnobMan:

<img width="541" alt="Screenshot 2024-01-17 at 20 59 52" src="https://github.com/RomanKubiak/ctrlr/assets/11859363/78047bed-37ae-492d-a8f1-636c02f8195b">

But I applaud fully the efforts of @damiensellier of implementing L&F V4 and indeed  proper VST3 export! What advantage would Juce 7 have over Juce 6?

**Reply by damiensellier** [reply]
wow your panel is looking great!
Same, I always use knobman it's a great tool.
I don't see any real benefits in juce 7 over 6 since we know Mac silicon can run native juce 6 programs.

**Reply by bijlevel** [reply]
@damiensellier Thanks! 
I still have to try and export an AU/VST3 for a panel based on the build of a Silicon CTRLR AU/VST3 with Juce 6. If that succeeds, then there's indeed no need for Juce 7. I will let you know

**Reply by damiensellier** [reply]
I have everything working great, with the global lookandfeel property changing the V1 V2 V3 or V4 style.
You will be able to override the LookAndFeel version for each component and mix them all for more crazyness.
I'm trying to figure some way to make it fully retrocompatible. It will be probably with a version tag for everything before this version use LookAndFeel_V3 and the newer V4 as default or i'll put everything to V3 as default, it's not a big deal.

**Reply by bijlevel** [reply]
@damiensellier Congratulations! I wil asap download the source and compile it on my computer and of course will let you know of my findings. The default L&F version is not a big deal, if users can switch easily between versions and even mix them. That would result indeed in some 'different' GUI's ;-)! (BTW still the build AU and VST3 on my Silicon Mac do fail to load into my DAW, but I think this is more a discussion for #565.)

**Reply by damiensellier** [reply]
@bijlevel sure you'll let me know. As soon as i have it 100% ready, I'll post a link to the source files and will join an intel mac version of it here as standalone VST3 and AU.
I can already see some nice possibilities as we could sselect color templates for sliders.
That would be great to build LookAndFeel classes for different components and give the users even more styling possibilities.

**Reply by damiensellier** [reply]
Tonight I worked on the Colour Picker, this is very handy  : 

![Capture d’écran, le 2024-01-18 à 23 16 34](https://github.com/RomanKubiak/ctrlr/assets/55164385/7d972309-9788-4c1f-a15b-a9f74e84e8be)
